src/include/duckdb/common/single_thread_pointer.h

```diff
@@ -0,0 +1,169 @@
+#include <cstdlib>
+
+template <typename T>
+class single_thread_ptr {
+public:
+	T *ptr;              // contained pointer
+	uint32_t *ref_count; // reference counter
+
+public:
+	// Default constructor, constructs an empty single_thread_ptr.
+	constexpr single_thread_ptr() : ptr(nullptr), ref_count(nullptr) {
+	}
+	// Construct empty single_thread_ptr.
+	constexpr single_thread_ptr(std::nullptr_t) : ptr(nullptr), ref_count(nullptr) {
+	}
+	// Construct a single_thread_ptr that wraps raw pointer.
+	explicit single_thread_ptr(T *p) {
+		if (p) {
+			ptr = p;
+			ref_count = new uint32_t(1);
+		}
+	}
+
+	single_thread_ptr(uint32_t *r, T *p) {
+		ptr = p;
+		ref_count = r;
+	}
+
+	template <class U>
+	single_thread_ptr(uint32_t *r, U *p) {
+		ptr = p;
+		ref_count = r;
+	}
+
+	// Construct a single_thread_ptr that wraps raw pointer of convertible type.
+	template <typename U>
+	single_thread_ptr(U *p) {
+		if (p) {
+			ptr = p;
+			ref_count = new uint32_t(1);
+		}
+	}
+
+	// Copy  constructor.
+	single_thread_ptr(const single_thread_ptr &sp) : ptr(nullptr), ref_count(nullptr) {
+		if (sp.ptr) {
+			ptr = sp.ptr;
+			ref_count = sp.ref_count;
+			++(*ref_count);
+		}
+	}
+
+	// Conversion constructor.
+	template <typename U>
+	single_thread_ptr(const single_thread_ptr<U> &sp) : ptr(nullptr), ref_count(nullptr) {
+		if (sp.ptr) {
+			ptr = sp.ptr;
+			ref_count = sp.ref_count;
+			++(*ref_count);
+		}
+	}
+
+	// move  constructor.
+	single_thread_ptr(single_thread_ptr &&sp) noexcept : ptr {sp.ptr}, ref_count {sp.ref_count} {
+		sp.ptr = nullptr;
+		sp.ref_count = nullptr;
+	}
+
+	// move  constructor.
+	template <class U>
+	single_thread_ptr(single_thread_ptr<U> &&sp) noexcept : ptr {sp.ptr}, ref_count {sp.ref_count} {
+		sp.ptr = nullptr;
+		sp.ref_count = nullptr;
+	}
+
+	// No effect if single_thread_ptr is empty or use_count() > 1, otherwise release the resources.
+	~single_thread_ptr() {
+		release();
+	}
+
+	void release() {
+		if (ptr && ref_count) {
+			if (--(*ref_count) == 0) {
+				delete ptr;
+			}
+		}
+		ptr = nullptr;
+	}
+
+	// Copy assignment.
+	single_thread_ptr &operator=(single_thread_ptr sp) noexcept {
+		std::swap(this->ptr, sp.ptr);
+		std::swap(this->ref_count, sp.ref_count);
+		return *this;
+	}
+
+	// Dereference pointer to managed object.
+	T &operator*() const noexcept {
+		return *ptr;
+	}
+	T *operator->() const noexcept {
+		return ptr;
+	}
+
+	// Return the contained pointer.
+	T *get() const noexcept {
+		return ptr;
+	}
+
+	// Return use count (use count == 0 if single_thread_ptr is empty).
+	long use_count() const noexcept {
+		if (ptr)
+			return *ref_count;
+		else
+			return 0;
+	}
+
+	// Check if there is an associated managed object.
+	explicit operator bool() const noexcept {
+		return (ptr);
+	}
+
+	// Resets single_thread_ptr to empty.
+	void reset() noexcept {
+		release();
+	}
+};
+
+template <class T>
+struct _object_and_block {
+	T object;
+	uint32_t pn = 1;
```

Comment:
Maybe put the count first, since accessing an object will always first access the count and then the object?

Reply:
Sure!